packages/camera/camera_avfoundation/ios/Classes/CameraPlugin.m

```diff
@@ -148,274 +138,355 @@ - (void)availableCamerasWithCompletion:
-                                    resolutionPreset:resolutionPreset
-                                       mediaSettings:mediaSettings
-                              mediaSettingsAVWrapper:mediaSettingsAVWrapper
-                                         orientation:[[UIDevice currentDevice] orientation]
-                                 captureSessionQueue:strongSelf.captureSessionQueue
-                                               error:&error];
-
-    if (error) {
-      result(FlutterErrorFromNSError(error));
-    } else {
-      if (strongSelf.camera) {
-        [strongSelf.camera close];
-      }
-      strongSelf.camera = cam;
-      [strongSelf.registry registerTexture:cam
-                                completion:^(int64_t textureId) {
-                                  result(@{
-                                    @"cameraId" : @(textureId),
-                                  });
-                                }];
+- (void)lockCaptureOrientation:(FCPPlatformDeviceOrientation)orientation
+                    completion:(nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera lockCaptureOrientation:orientation];
+    completion(nil);
+  });
+}
+
+- (void)unlockCaptureOrientationWithCompletion:
+    (nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera unlockCaptureOrientation];
+    completion(nil);
+  });
+}
+
+- (void)pausePreviewWithCompletion:(nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera pausePreview];
+    completion(nil);
+  });
+}
+
+- (void)resumePreviewWithCompletion:(nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera resumePreview];
+    completion(nil);
+  });
+}
+
+- (void)setImageFileFormat:(FCPPlatformImageFileFormat)format
+                completion:(nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera setImageFileFormat:format];
+    completion(nil);
+  });
+}
+
+- (void)updateDescriptionWhileRecordingCameraName:(nonnull NSString *)cameraName
+                                       completion:
+                                           (nonnull void (^)(FlutterError *_Nullable))completion {
+  __weak typeof(self) weakSelf = self;
+  dispatch_async(self.captureSessionQueue, ^{
+    [weakSelf.camera setDescriptionWhileRecording:cameraName withCompletion:completion];
+  });
+}
+
+- (void)disposeCamera:(NSInteger)cameraId
+           completion:(nonnull void (^)(FlutterError *_Nullable))completion {
+  [_registry unregisterTexture:cameraId];
```

Comment:
This is what made me remove the thread abstraction around the registry—this was one of two meaningful uses of the wrapper, and it was wrong. The old code was calling this in the async handler, which means it was effectively in the `captureSessionQueue` dispatch block below. But it was internally moving the call back to the main thread, which while correct, *was not awaited* before continuing. That meant that unregistering the texture—which ensures that the engine won't keep asking for textures—was not reliably being done *before* tearing down the camera on the session queue.

It may have worked out later due to the way thread redispatch is done when actually doing the copy, but I didn't try to unwind that since it's clearly better to do in this order.

Reply:
good call!